pymatgen/core/surface.py

```diff
@@ -961,126 +1037,140 @@ def get_slab(self, shift=0, tol: float = 0.1, energy=None):
+                    dist_matrix[i, j] = z_dist
+                    dist_matrix[j, i] = z_dist
+
+            # Cluster the sites by z coordinates
+            z_matrix = linkage(squareform(dist_matrix))
+            clusters = fcluster(z_matrix, ftol, criterion="distance")
+
+            # Generate a cluster to z coordinate mapping
+            clst_loc = {c: frac_coords[i][2] for i, c in enumerate(clusters)}
+
+            # Wrap all clusters into the unit cell ([0, 1) range)
+            possible_clst = [coord - math.floor(coord) for coord in sorted(clst_loc.values())]
+
+            # Calculate shifts
+            n_shifts = len(possible_clst)
+            shifts = []
+            for i in range(n_shifts):
+                # Handle the special case for the first-last
+                # z coordinate (because of periodic boundary condition)
+                if i == n_shifts - 1:
+                    # TODO (@DanielYang59): Why calculate the "center" of the
+                    # two clusters, which is not actually the shift?
+                    shift = (possible_clst[0] + 1 + possible_clst[i]) * 0.5
+
+                else:
+                    shift = (possible_clst[i] + possible_clst[i + 1]) * 0.5
+
+                shifts.append(shift - math.floor(shift))
+
+            return sorted(shifts)
+
+        def get_z_ranges(bonds: dict[tuple[Species, Species], float], tol: float) -> list[tuple[float, float]]:
+            """Calculate list of z ranges where each z_range is a (lower_z, upper_z) tuple.
+
+            Args:
+                bonds (dict): specified as a (specie1, specie2): max_bond_dist dict.
+                tol (float): Tolerance for determine overlapping positions.
+            """
+            # Sanitize species in dict keys
+            bonds = {(get_el_sp(s1), get_el_sp(s2)): dist for (s1, s2), dist in bonds.items()}
+
+            z_ranges = []
+            for (sp1, sp2), bond_dist in bonds.items():
+                for site in self.oriented_unit_cell:
+                    if sp1 in site.species:
+                        for nn in self.oriented_unit_cell.get_neighbors(site, bond_dist):
+                            if sp2 in nn.species:
+                                z_range = tuple(sorted([site.frac_coords[2], nn.frac_coords[2]]))
+
+                                # Handle cases when z coordinate of site goes
+                                # beyond the upper boundary
+                                if z_range[1] > 1:
+                                    z_ranges.extend([(z_range[0], 1), (0, z_range[1] - 1)])
+
+                                # When z coordinate is below the lower boundary
+                                elif z_range[0] < 0:
+                                    z_ranges.extend([(0, z_range[1]), (z_range[0] + 1, 1)])
+
+                                # Neglect overlapping positions
+                                elif z_range[0] != z_range[1]:
```

Comment:
Using `==` for detecting equality of z coordinates seems improper, but replacing it with `isclose(z_range[0], z_range[1], abs_tol=tol)` breaks some unit tests. Need a closer look.